network/hybridNetwork_test.go

```diff
@@ -0,0 +1,143 @@
+	relayCfg.NetAddress = addr
+	relayCfg.PublicAddress = addr
+	netA, err = NewHybridP2PNetwork(log.With("node", "netA"), relayCfg, p2pKeyDir, nil, genesisID, "net", &nopeNodeInfo{})
+	require.NoError(t, err)
+
+	err = netA.Start()
+	require.NoError(t, err)
+	defer netA.Stop()
+
+	// collect relay address and prepare nodes phonebook
+	peerInfoA := netA.p2pNetwork.service.AddrInfo()
+	addrsAp2p, err := peer.AddrInfoToP2pAddrs(&peerInfoA)
+	require.NoError(t, err)
+	require.NotZero(t, addrsAp2p[0])
+	multiAddrStr := addrsAp2p[0].String()
+
+	fullAddr, portListen := netA.wsNetwork.Address()
+	require.True(t, portListen)
+	require.NotZero(t, addr)
+	require.Contains(t, fullAddr, addr)
+
+	phoneBookAddresses := []string{multiAddrStr, addr}
+
+	netB, err := NewHybridP2PNetwork(log.With("node", "netB"), cfg, "", phoneBookAddresses, genesisID, "net", &nopeNodeInfo{})
+	require.NoError(t, err)
+	// for netB start the p2p network first
+	err = netB.p2pNetwork.Start()
+	require.NoError(t, err)
+	defer netB.Stop()
+
+	netC, err := NewHybridP2PNetwork(log.With("node", "netC"), cfg, "", phoneBookAddresses, genesisID, "net", &nopeNodeInfo{})
+	require.NoError(t, err)
+	// for netC start the ws network first
+	err = netC.wsNetwork.Start()
+	require.NoError(t, err)
+	defer netC.Stop()
+
+	// ensure initial connections are done
+	require.Eventually(t, func() bool {
+		return len(netA.GetPeers(PeersConnectedIn)) == 2
+	}, 3*time.Second, 50*time.Millisecond)
+
+	// start the second half of the hybrid net
+	err = netB.wsNetwork.Start()
+	require.NoError(t, err)
+	err = netC.p2pNetwork.Start()
+	require.NoError(t, err)
+
+	// wait for connection attempts. nodes need some time to make connections,
+	// and instead of `time.Sleep(1 * time.Second)` the networkPeerIdentityDisconnect net identity counter is used.
+	// Since this test is not parallel the networkPeerIdentityDisconnect should not be modified from outside.
+	require.Eventually(t, func() bool {
+		return networkPeerIdentityDisconnect.GetUint64Value() == 2+identDiscValue
+	}, 3*time.Second, 50*time.Millisecond,
+		"disconnect counter: %d, expected: %d", networkPeerIdentityDisconnect.GetUint64Value(), 2+identDiscValue)
+
+	// now count connections
+	// netA should have 2 connections, not 4
+	// netB should have 1 connection (via p2p)
+	// netC should have 1 connection (via ws)
+
+	require.Eventually(t, func() bool {
+		return len(netB.GetPeers(PeersConnectedOut)) == 1
+	}, 3*time.Second, 50*time.Millisecond,
+		"netA in: %d, netB out: %d, netC out: %d",
+		len(netA.GetPeers(PeersConnectedIn)), len(netB.GetPeers(PeersConnectedOut)), len(netC.GetPeers(PeersConnectedOut)))
+
+	require.Eventually(t, func() bool {
+		return len(netC.GetPeers(PeersConnectedOut)) == 1
+	}, 3*time.Second, 50*time.Millisecond,
+		"netA in: %d, netB out: %d, netC out: %d",
+		len(netA.GetPeers(PeersConnectedIn)), len(netB.GetPeers(PeersConnectedOut)), len(netC.GetPeers(PeersConnectedOut)))
+
+	require.Eventually(t, func() bool {
```

Comment:
the [last failure](https://app.circleci.com/pipelines/github/algorand/go-algorand/18585/workflows/b2cf691e-4d1d-4228-b5ae-15f5694da29b/jobs/274194) is interesting:
```
        	Test:       	TestHybridNetwork_DuplicateConn
        	Messages:   	netA in: 3, netB out: 1, netC out: 1
```